component/loki/process/stages/eventlogmessage.go

```diff
@@ -0,0 +1,134 @@
+package stages
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	defaultSource = "message"
+)
+
+type EventLogMessageConfig struct {
+	Source            string `river:"source,attr,optional"`
+	DropInvalidLabels bool   `river:"drop_invalid_labels,attr,optional"`
+	OverwriteExisting bool   `river:"overwrite_existing,attr,optional"`
+}
+
+func (e *EventLogMessageConfig) Validate() error {
+	if !model.LabelName(e.Source).IsValid() {
+		return fmt.Errorf(ErrInvalidLabelName, e.Source)
+	}
+	return nil
+}
+
+func (e *EventLogMessageConfig) SetToDefault() {
+	e.Source = defaultSource
+}
+
+type eventLogMessageStage struct {
+	cfg    *EventLogMessageConfig
+	logger log.Logger
+}
+
+// Create a event log message stage, including validating any supplied configuration
+func newEventLogMessageStage(logger log.Logger, cfg *EventLogMessageConfig) Stage {
+	return &eventLogMessageStage{
+		cfg:    cfg,
+		logger: log.With(logger, "component", "stage", "type", "event_log_message"),
+	}
+}
+
+func (m *eventLogMessageStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	key := m.cfg.Source
+	go func() {
+		defer close(out)
+		for e := range in {
+			err := m.processEntry(e.Extracted, key)
+			if err != nil {
+				continue
+			}
+			out <- e
+		}
+	}()
+	return out
+}
+
+// Process a event log message from extracted with the specified key, adding additional
+// entries into the extracted map
+func (m *eventLogMessageStage) processEntry(extracted map[string]interface{}, key string) error {
+	value, ok := extracted[key]
+	if !ok {
+		if Debug {
```

Comment:
When we initially ported over the Loki components we made an effort to stop using this pattern.

After a while, we saw users report _significant_ CPU % increase out of the blue, and turns out it came from debug logs in busy logs pipelines. 

Let's keep it as a discussion point for the future, but for now, let's remove this to allow debug-level logs to work properly (since IIRC there's no way wired up to set Debug to true in Flow mode).

Reply:
I think that @rfratto wanted to use this new behaviour actually: https://github.com/grafana/agent/issues/5060#issuecomment-1710349195 

And I can see why - we have reports of CPU usage being a problem.

To solve this I think we should use a different approach altogether. I think the issue should be solved by getting rid of `Log(fmt.Sprintf(...))` and instead use `Log("msg", <message>, "sth", <sth>, ...)` so that the call to `Log` function doesn't require to build a new string. And then the very first thing that our loggers should do is to check for `level` and skip the log line if it's below the logger's level. That way the log function can be inlined by compiler and we don't need to surround every call with `if Debug { ... }`. I believe this is how performant logging frameworks do it.

Reply:
Yeah, let's not block this just for the debug statement.